waspc/data/Generator/templates/server/src/auth/providers/oauth/createRouter.ts

```diff
@@ -43,47 +46,43 @@ export function createRouter(provider: ProviderConfig, initData: { passportStrat
           }
 
           // Wrap call to getUserFieldsFn so we can invoke only if needed.
-          const getUserFields = () => getUserFieldsFn(contextWithUserEntity, { profile: providerProfile });
+          const getUserFields = () => getUserFieldsFn ? getUserFieldsFn(contextWithUserEntity, { profile: providerProfile }) : Promise.resolve({});
           // TODO: In the future we could make this configurable, possibly associating an external account
           // with the currently logged in account, or by some DB lookup.
-          const user = await findOrCreateUserByExternalAuthAssociation(provider.id, providerProfile.id, getUserFields);
+          const auth = await findOrCreateAuthByAuthIdentity(provider.id, providerProfile.id, getUserFields);
 
-          const token = await sign(user.id);
+          const token = await sign(auth.{= userFieldOnAuthEntityName =}.id);
           res.json({ token });
       })
     )
 
     return router;
 }
 
-async function findOrCreateUserByExternalAuthAssociation(
-  provider: string,
-  providerId: string,
+async function findOrCreateAuthByAuthIdentity(
+  providerName: string,
+  providerUserId: string,
   getUserFields: () => ReturnType<GetUserFieldsFn>,
-): Promise<{= userEntityUpper =}> {
+) {
   // Attempt to find a User by an external auth association.
-  const externalAuthAssociation = await prisma.{= externalAuthEntityLower =}.findFirst({
-    where: { provider, providerId },
-    include: { user: true }
+  const authIdentity = await prisma.{= authIdentityEntityLower =}.findFirst({
+    where: { providerName, providerUserId },
+    include: {
+      {= authFieldOnAuthIdentityEntityName =}: {
+        include: {
+          {= userFieldOnAuthEntityName =}: true
+        }
+      }
+    }
   })
 
-  if (externalAuthAssociation) {
-    return externalAuthAssociation.user
+  if (authIdentity) {
+    return authIdentity.{= authFieldOnAuthIdentityEntityName =}
   }
 
-  // No external auth association linkage found. Create a new User using details from
-  // `getUserFields()`. Additionally, associate the externalAuthAssociations with the new User.
   const userFields = await getUserFields()
-  const userAndExternalAuthAssociation = {
-    ...userFields,
-    {=# isPasswordOnUserEntity =}
-    // TODO: Decouple social from usernameAndPassword auth.
-    password: uuidv4(),
-    {=/ isPasswordOnUserEntity =}
-    externalAuthAssociations: {
-      create: [{ provider, providerId }]
-    }
-  }
 
-  return createUser(userAndExternalAuthAssociation)
+  const auth = await createAuthWithUser(providerName, providerUserId, undefined, userFields)
+  // NOTE: we are fetching the auth again becuase it incldues nested user
```

Comment:
Can't we then just remove nested user from it and return it instead of fetching it again?

Reply:
I might have worded the comment in a poor way, the object we have doesn't have all the fields we need.

I'll examine this and see if we can avoid fetching the user again 👍

Reply:
Hm ok I am still not 100% what you wanted to do but you can reword the comment to make it clearer.

Reply:
I've refactored this bit to not have extra DB calls 🚀